tvm/src/schedule/schedule_dataflow_rewrite.cc

```diff
@@ -453,84 +393,60 @@ void Schedule::stage_move(Stage parent, DeviceType device_type,
                          op->input_placeholders, op->output_placeholders, body);
 }
 
-// annotate the tensor to be joined
-void Schedule::join_to(const Tensor& target, Stage source, Stage dest,
-                       StreamType stream_type, int channel_depth) {
+// Move data to device
+Tensor Schedule::move_to(const Tensor& target, Stage parent,
+                         DeviceType device_type, StreamType stream_type,
+                         int channel_depth, Array<Expr> dev_ports) {
   Stage target_stage = (*this)[target];
+  std::vector<Stage> consumers;
   size_t num_stage = (*this)->stages.size();
   Buffer target_buffer;
 
-  const PlaceholderOpNode* op = target_stage->op.as<PlaceholderOpNode>();
-  bool is_placeholder = op ? true : false;
-  if (is_placeholder) {
-    for (size_t i = 0; i < num_stage; i++) {
-      Stage s = (*this)->stages[i];
-      if (const ExternOpNode* op = s->op.as<ExternOpNode>()) {
-        for (size_t j = 0; j < op->inputs.size(); j++) {
-          if (target == op->inputs[j]) {
-            target_buffer = op->input_placeholders[j];
-          }
-        }
-      }
+  // parse the memory module interface
+  CHECK_EQ(dev_ports.size(), 3);
+  auto mem_type = dev_ports[0].as<IntImm>()->value;
+  StorageType storage = static_cast<StorageType>(mem_type);
+
+  // Bind IO inertface to on-chip storage
+  if (storage == StorageType::devBRAM || storage == StorageType::devLUTRAM ||
+      storage == StorageType::devURAM) {
+    std::string private_dev;
+    if (storage == StorageType::devBRAM) {
+      private_dev = "BRAM";
+    } else if (storage == StorageType::devLUTRAM) {
+      private_dev = "LUTRAM";
+    } else {
+      private_dev = "URAM";
     }
-  } else {  // mark device scope of consumers & update kernel stmts
-    const ExternOpNode* op = target_stage->op.as<ExternOpNode>();
-    target_buffer = op->output_placeholders[0];
-  }
-
-  CHECK(source.defined());
-  const ExternOpNode* src_op = source->op.as<ExternOpNode>();
-  CHECK(src_op) << "cannot join placeholder stage " << source;
 
-  InfoUpdater::channelCount += 1;
-  auto index = InfoUpdater::channelCount;
+    auto binding = dev_ports[1];
+    HCL_DEBUG_LEVEL(2) << "[debug] Assign tensor " << target << " to "
+                       << private_dev << " with attr " << binding;
```

Comment:
It seems `private_dev` if only used for printing? In this case, I guess we don't need these.

Reply:
Removed

Reply:
Wait. I found this is actually implemented. The function body is mutated and a new AttrStmt is added to the IR
```c++
    Stmt body = AttrStmt::make(node, attr::bind_scope, binding, op->body);
    target_stage->op = ExternOpNode::make(op->name, op->tag, op->axis,
                                          op->inputs, op->input_placeholders,
                                          op->output_placeholders, **op->body);**
```